File: day-15/day-15.r
# advent of code
# day 15

# part 1
gameNumbers <- list(
  '9' = 1,
  '12' = 2,
  '1' = 3,
  '4' = 4,
  '17' = 5,
  '0' = 6,
  '18' = 7
)
last <- 18

nextGameNumber <- function(last, gameNumbers = gameNumbers){
  if(length(gameNumbers[[as.character(last)]]) == 1){
    return(0)
  } else {
    gameNumbers[[as.character(last)]] <- rev(rev(gameNumbers[[as.character(last)]])[1:2])
    lastTimes <- rev(gameNumbers[[as.character(last)]])
    return(lastTimes[1] - lastTimes[2])
  }
}

for(i in 8:30000000){
  if(i %% 10000 == 0){print(i)}
  nextNumber <- nextGameNumber(last, gameNumbers)
  gameNumbers[[as.character(nextNumber)]] <- c(gameNumbers[[as.character(nextNumber)]], i)
  last <- nextNumber
}

File: day-15/day-15.py
# advent of code
# day 15

# part 1
import sys
import time

input = [9, 12, 1, 4, 17, 0, 18]

def play(end):
    lastPlayed = {}
    for i in range(len(input) - 1):
        lastPlayed[input[i]] = i
    next = input[-1]
    for i in range(len(input), end + 1):
        last = next
        if(last in lastPlayed.keys()):
            next = i - 1 - lastPlayed[last]
            lastPlayed[last] = i - 1
        else:
            lastPlayed[last] = i - 1
            next = 0
    print(last)

start_time = time.time()
play(2020)
end_time = time.time()
print('%s seconds' % (end_time - start_time))

# part 2
start_time = time.time()
play(30000000)
end_time = time.time()
print('%s seconds' % (end_time - start_time))
